x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java

```diff
@@ -249,41 +248,36 @@ private void ensureOpen() {
     }
 
     CompletableFuture<Integer> fetchRange(
-        long position,
+        Tuple<Long, Long> range,
         CheckedBiFunction<Long, Long, Integer, IOException> onRangeAvailable,
         CheckedBiConsumer<Long, Long, IOException> onRangeMissing
     ) {
         final CompletableFuture<Integer> future = new CompletableFuture<>();
         try {
-            if (position < 0 || position > tracker.getLength()) {
-                throw new IllegalArgumentException("Wrong read position [" + position + "]");
+            if (range.v1() < 0 || range.v1() > tracker.getLength() || range.v2() < 0 || range.v2() > tracker.getLength()) {
+                throw new IllegalArgumentException(
+                    "Invalid range [start=" + range.v1() + ", end=" + range.v2() + "] for length [" + tracker.getLength() + ']'
+                );
             }
-
             ensureOpen();
-            final long rangeStart = (position / rangeSize) * rangeSize;
-            final long rangeEnd = Math.min(rangeStart + rangeSize, tracker.getLength());
-
             final List<SparseFileTracker.Gap> gaps = tracker.waitForRange(
-                rangeStart,
-                rangeEnd,
+                range.v1(),
+                range.v2(),
                 ActionListener.wrap(
-                    rangeReady -> future.complete(onRangeAvailable.apply(rangeStart, rangeEnd)),
+                    rangeReady -> future.complete(onRangeAvailable.apply(range.v1(), range.v2())),
                     rangeFailure -> future.completeExceptionally(rangeFailure)
                 )
             );
 
             if (gaps.size() > 0) {
```

Comment:
Don't really need this guard any more.

Reply:
Removed in e9831a3c1c5f2035586b08bf390667a149313359